File: solutions/fsharp/cars-assemble/2/CarsAssemble.fs
module CarsAssemble

let CARS_PER_HOUR_PER_SPEED = 221

let successRate (speed: int) : float =
    match speed with
    | 0 -> 0.0
    | s when s < 5 -> 1.0
    | s when s < 9 -> 0.9
    | 9 -> 0.8
    | 10 -> 0.77
    | _ -> failwith "invalid speed"

let productionRatePerHour (speed: int) : float =
    float speed * float CARS_PER_HOUR_PER_SPEED * (successRate speed)

let workingItemsPerMinute (speed: int) : int =
    int ((productionRatePerHour speed) / float 60)

File: solutions/csharp/cars-assemble/2/CarsAssemble.cs
static class AssemblyLine
{
  const int CarsPerHour = 221;

  public static double SuccessRate(int speed)
  {
    if (speed == 0)
      return 0;
    else if (speed <= 4)
      return 1;
    else if (speed <= 8)
      return .9;
    else if (speed == 9)
      return .8;
    return .77;
  }

  public static double ProductionRatePerHour(int speed) =>
    speed * CarsPerHour * SuccessRate(speed);

  public static int WorkingItemsPerMinute(int speed) =>
    (int)ProductionRatePerHour(speed) / 60;
}
